evm/src/cpu/kernel/tests/block_hash.rs

```diff
@@ -0,0 +1,97 @@
+
+    let block_number: u8 = rng.gen();
+    let initial_stack = vec![retdest, block_number.into()];
+
+    let hashes: Vec<U256> = vec![U256::from_big_endian(&thread_rng().gen::<H256>().0); 257];
+
+    let mut interpreter = Interpreter::new_with_kernel(blockhash_label, initial_stack);
+    interpreter.set_memory_segment(Segment::BlockHashes, hashes[0..256].to_vec());
+    interpreter.set_global_metadata_field(GlobalMetadata::BlockCurrentHash, hashes[256]);
+    interpreter.set_global_metadata_field(GlobalMetadata::BlockNumber, 256.into());
+    interpreter.run()?;
+
+    let result = interpreter.stack();
+    assert_eq!(
+        result[0], hashes[block_number as usize],
+        "Resulting block hash {:?} different from expected hash {:?}",
+        result[0], hashes[block_number as usize]
+    );
+
+    Ok(())
+}
+
+#[test]
+fn test_big_index_block_hash() -> Result<()> {
+    let mut rng = rand::thread_rng();
+
+    let blockhash_label = KERNEL.global_labels["blockhash"];
+    let retdest = 0xDEADBEEFu32.into();
+    let cur_block_number = 3;
+    let block_number: usize = rng.gen::<u8>() as usize;
+    let actual_block_number = block_number + cur_block_number;
+    let initial_stack = vec![retdest, actual_block_number.into()];
+
+    let hashes: Vec<U256> = vec![U256::from_big_endian(&thread_rng().gen::<H256>().0); 257];
+
+    let mut interpreter = Interpreter::new_with_kernel(blockhash_label, initial_stack);
+    interpreter.set_memory_segment(Segment::BlockHashes, hashes[0..256].to_vec());
+    interpreter.set_global_metadata_field(GlobalMetadata::BlockCurrentHash, hashes[256]);
+    interpreter.set_global_metadata_field(GlobalMetadata::BlockNumber, cur_block_number.into());
+    interpreter.run()?;
+
+    let result = interpreter.stack();
+    assert_eq!(
+        result[0],
+        0.into(),
+        "Resulting block hash {:?} different from expected hash {:?}",
+        result[0],
+        0
+    );
+
+    Ok(())
+}
+
+#[test]
+fn test_small_index_block_hash() -> Result<()> {
+    let mut rng = rand::thread_rng();
+
+    let blockhash_label = KERNEL.global_labels["blockhash"];
+    let retdest = 0xDEADBEEFu32.into();
+    let cur_block_number = 512;
+    let block_number = rng.gen::<u8>() as usize;
+    let initial_stack = vec![retdest, block_number.into()];
+
+    let hashes: Vec<U256> = (20..277).map(|elt| elt.into()).collect();
+
+    let mut interpreter = Interpreter::new_with_kernel(blockhash_label, initial_stack);
+    interpreter.set_memory_segment(Segment::BlockHashes, hashes[0..256].to_vec());
+    interpreter.set_global_metadata_field(GlobalMetadata::BlockCurrentHash, hashes[256]);
+    interpreter.set_global_metadata_field(GlobalMetadata::BlockNumber, cur_block_number.into());
+    interpreter.run()?;
+
+    let result = interpreter.stack();
+    assert_eq!(
+        result[0],
+        0.into(),
+        "Resulting block hash {:?} different from expected hash {:?}",
+        result[0],
+        0
+    );
+
+    Ok(())
+}
```

Comment:
Could be nice to have a test for the overflow, e.g. using `cur_block_number=0` and `block_number =  U256::MAX`.

Reply:
When I added the overflow test, the interpreter failed with 'Integer overflow when casting to usize' in `recert_checkpoint`. So I'll add the test with `#[should_panic]` on top.

Reply:
See my other comment, if we remove `%add_or_fault` this shouldn't panic.